src/Parsers/TokenIterator.cpp

```diff
@@ -20,13 +21,59 @@ Tokens::Tokens(const char * begin, const char * end, size_t max_query_size, bool
 
 UnmatchedParentheses checkUnmatchedParentheses(TokenIterator begin)
 {
+    std::unordered_set<String> valid_kql_negative_suffix(
+        {
+         "between",
+         "contains",
+         "contains_cs",
+         "endswith",
+         "endswith_cs",
+         "~",
+         "=",
+         "has",
+         "has_cs",
+         "hasprefix",
+         "hasprefix_cs",
+         "hassuffix",
+         "hassuffix_cs",
+         "in",
+         "startswith",
+         "startswith_cs"});
     /// We have just two kind of parentheses: () and [].
     UnmatchedParentheses stack;
 
     /// We have to iterate through all tokens until the end to avoid false positive "Unmatched parentheses" error
     /// when parser failed in the middle of the query.
-    for (TokenIterator it = begin; it.isValid(); ++it)
+    for (TokenIterator it = begin; !it->isEnd(); ++it)
     {
+        if (!it.isValid()) // allow kql negative operators
+        {
+            if (it->type == TokenType::ErrorSingleExclamationMark)
+            {
+                ++it;
+                if (!valid_kql_negative_suffix.contains(String(it.get().begin, it.get().end)))
+                    break;
+                --it;
+            }
+            else if (it->type == TokenType::ErrorWrongNumber)
+            {
+                if (!ParserKQLDateTypeTimespan().parseConstKQLTimespan(String(it.get().begin, it.get().end)))
+                    break;
+            }
+            else
+            {
+                if (String(it.get().begin, it.get().end) == "~")
+                {
+                    --it;
+                    if (const auto prev = String(it.get().begin, it.get().end); prev != "!" && prev != "=" && prev != "in")
+                        break;
+                    ++it;
+                }
+                else
+                    break;
+            }
+        }
+
```

Comment:
I will check this

Reply:
Hi @yakov-olkhovskiy , this piece of code is to enable the negative operators inside the `embed kql table function`.
because its' a part of regular SQL query statement, a separated specific tokenizer won't help in this case.
I have changed it to only affect  the tokens inside the `kql table function` , without affecting normal SQL, for example:

`select * from kql(Customers | where FirstName !contains_cs 'Pet');  `
is allowed.

while `select * from Customers  where FirstName !contains_cs 'Pet'`  is not allowed as before:

```
Syntax error: failed at position 42 ('!'):

select * from Customers  where FirstName !contains_cs 'Pet'

Exclamation mark can only occur in != operator: '!'
```
can you check again ?

Thanks

Reply:
@yakov-olkhovskiy ,

Did get chance to review my comments? 

Thanks.